synapse/crypto/keyring.py

```diff
@@ -252,321 +251,225 @@ def verify_events_for_server(
+        # keys and the signature matches.
+        for key_id in verify_request.key_ids:
+            key_result = found_keys.get(key_id)
+            if not key_result:
+                raise SynapseError(
+                    401,
+                    f"Missing keys for {verify_request.server_name}: {key_id}",
+                    Codes.UNAUTHORIZED,
+                )
 
-        results = [process(r) for r in verify_requests]
+            if key_result.valid_until_ts < verify_request.minimum_valid_until_ts:
+                raise SynapseError(
+                    401,
+                    f"Failed to find key with recent enough `valid_until_ts` for {verify_request.server_name}: {key_id}",
+                    Codes.UNAUTHORIZED,
+                )
 
-        if key_lookups:
-            run_in_background(self._start_key_lookups, key_lookups)
+            verify_key = key_result.verify_key
+            try:
+                json_object = verify_request.get_json_object()
+                verify_signed_json(
+                    json_object,
+                    verify_request.server_name,
+                    verify_key,
+                )
+            except SignatureVerifyException as e:
+                logger.debug(
+                    "Error verifying signature for %s:%s:%s with key %s: %s",
+                    verify_request.server_name,
+                    verify_key.alg,
+                    verify_key.version,
+                    encode_verify_key_base64(verify_key),
+                    str(e),
+                )
+                raise SynapseError(
+                    401,
+                    "Invalid signature for server %s with key %s:%s: %s"
+                    % (
+                        verify_request.server_name,
+                        verify_key.alg,
+                        verify_key.version,
+                        str(e),
+                    ),
+                    Codes.UNAUTHORIZED,
+                )
 
-        return results
+    async def _inner_fetch_key_requests(
+        self, requests: List[_FetchKeyRequest]
+    ) -> Dict[str, Dict[str, FetchKeyResult]]:
+        """Processing function for the queue of `_FetchKeyRequest`."""
+
+        logger.debug("Starting fetch for %s", requests)
+
+        # First we need to deduplicate requests for the same key. We do this by
+        # taking the *maximum* requested `minimum_valid_until_ts` for each pair
+        # of server name/key ID.
+        server_to_key_to_ts = {}  # type: Dict[str, Dict[str, int]]
+        for request in requests:
+            by_server = server_to_key_to_ts.setdefault(request.server_name, {})
+            for key_id in request.key_ids:
+                existing_ts = by_server.get(key_id)
+
+                if existing_ts:
+                    by_server[key_id] = max(request.minimum_valid_until_ts, existing_ts)
+                else:
+                    by_server[key_id] = request.minimum_valid_until_ts
+
```

Comment:
```suggestion
                existing_ts = by_server.get(key_id, 0)
                by_server[key_id] = max(request.minimum_valid_until_ts, existing_ts)

```

or, if you prefer:

```suggestion
                existing_ts = by_server.get(key_id)

                if existing_ts is not None:
                    by_server[key_id] = max(request.minimum_valid_until_ts, existing_ts)
                else:
                    by_server[key_id] = request.minimum_valid_until_ts

```

right now it's neither one thing nor the other